File: articles/decision-science-in-financial-risk-management/julia/portfolio_loss_model.jl
# portfolio_loss_model.jl

function expected_loss(losses, probabilities)
    total = 0.0
    for i in eachindex(losses)
        total += losses[i] * probabilities[i]
    end
    return total
end

losses = [-1.2, -4.8, -3.6, -6.2]
probs = [0.55, 0.20, 0.15, 0.10]
println("Expected loss = ", round(expected_loss(losses, probs), digits=6))
println("Worst case = ", minimum(losses))

File: articles/decision-science-in-financial-risk-management/python/portfolio_loss_model.py
#!/usr/bin/env python3
"""Portfolio loss helper for financial risk decision science."""

from __future__ import annotations


def expected_loss(losses: dict[str, float], probabilities: dict[str, float]) -> float:
    return sum(losses[name] * probabilities[name] for name in losses)


def worst_case_loss(losses: dict[str, float]) -> float:
    return min(losses.values())


if __name__ == "__main__":
    losses = {"normal": -1.2, "recession": -4.8, "liquidity_shock": -3.6, "systemic_stress": -6.2}
    probs = {"normal": 0.55, "recession": 0.20, "liquidity_shock": 0.15, "systemic_stress": 0.10}
    print(round(expected_loss(losses, probs), 6))
    print(round(worst_case_loss(losses), 6))
